File: Java/Nth-to-Last-Node-in-List.java
/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: Nth to last node of a singly linked list. 
     */
    ListNode nthToLast(ListNode head, int n) {
        // write your code here
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode cur = dummy;
        int count = 0;
        while (count++ < n) {
        	cur = cur.next;
        }
        ListNode cnt = dummy;
        while (cur != null) {
        	cur = cur.next;
        	cnt = cnt.next;
        }
        return cnt;
    }
}

File: Cpp/Nth-to-Last-Node-in-List.cpp
/**
 * Definition of ListNode
 **/
class ListNode {
public:
    int val;
    ListNode *next;
    ListNode(int val) {
        this->val = val;
        this->next = NULL;
    }
};

class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: Nth to last node of a singly linked list.
     */
    ListNode *nthToLast(ListNode *head, int n) {
        ListNode *node = head;
        if (n == 0) return NULL;
        while (n - 1 > 0) {
            if (node == NULL) return NULL;
            node = node->next;
            n--;
        }
        while (node->next != NULL) {
            head = head->next;
            node = node->next;
        }
        return head;
        // write your code here
    }
};